main.c

```diff
@@ -0,0 +1,100 @@
+#include <signal.h>
+#include <stdio.h>
+
+#include "dns_sniffer.h"
+#include "firewall.h"
+
+#define IPV4 4
+#define IPV6 6
+
+#define NFLOG_GROUP 3
+
+#define SUCCESS_EXIT 0
+#define FAILURE_EXIT 1
+
+dns_sniffer_t sniffer;
+dns_callback_data_t callback_data;
+
+void cleanup() {
+    printf("Cleaning up...\n");
+    delete_incoming_dns_nflog_rule(IPV4, NFLOG_GROUP);
+    delete_incoming_dns_nflog_rule(IPV6, NFLOG_GROUP);
+
+    close_dns_sniffer(&sniffer);
+
+    if (callback_data.output_fd) {
+        fclose(callback_data.output_fd);
+        callback_data.output_fd = NULL;
+    }
+}
+
+void setup_iptables_rules() {
+    /* removing any existing rules from previous runs */
+    delete_incoming_dns_nflog_rule(IPV4, NFLOG_GROUP);
+    delete_incoming_dns_nflog_rule(IPV6, NFLOG_GROUP);
+
+    add_incoming_dns_nflog_rule(IPV4, NFLOG_GROUP);
+    add_incoming_dns_nflog_rule(IPV6, NFLOG_GROUP);
+}
+
+void cb_print_dns_packet(dns_response_t *response, FILE *output_fd) {
+    fprintf(output_fd, "Server: %s, Domain: %s, IP Version: %s, Query Type: %s\n",
+           response->dns_server,
+           response->domain,
+           (response->ip_version == IPV4) ? "IPv4" : "IPv6",
+           (response->query_type == A) ? "A" : (response->query_type == AAAA) ? "AAAA"
+                                           : (response->query_type == CNAME)  ? "CNAME"
+                                                                              : "Unknown");
+}
+
+void signal_handler(int signum) {
+    printf("got signal %d\n", signum);
+    cleanup();
```

Comment:
you are in signal handler, do minimal work, raise flag and let the program do the cleanup